activesupport/lib/active_support/message_pack/serializer.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module MessagePack
+    module Serializer # :nodoc:
+      attr_accessor :message_pack_factory
+
+      def dump(object)
+        @packer_key ||= "message_pack_packer_#{object_id}"
+        packer = (IsolatedExecutionState[@packer_key] ||= message_pack_factory.packer)
+
+        # TODO RFC:
```

Comment:
So a prefix is a good idea, but I'd heavily advise to have one byte for versioning, as if the format need to change in the future, that may save your bacon.

Reply:
Also it may be worth to make it a single magic Integer, so that you call `write` only once.

Reply:
My intention here was to allow consuming code to guess the serialization format, the way [`SerializedCookieJars` does](https://github.com/rails/rails/blob/89bd41201a060e8e6ee39591430d136e9e012c34/actionpack/lib/action_dispatch/middleware/cookies.rb#L571-L576) and [`JsonWithMarshalFallback` does](https://github.com/rails/rails/blob/89bd41201a060e8e6ee39591430d136e9e012c34/activesupport/lib/active_support/json_with_marshal_fallback.rb#L4-L31) for `Marshal`.  I think a single byte version number might make that more difficult.

For versioning, I was thinking there might be other options:

1. We could add a new type ID for the class while keeping the old ID to deserialize old messages, the way you did for `Time` in https://github.com/Shopify/paquito/commit/6ac845d7bf7fe4318f8ae310d7fe8de80fc5fc3c.

    It seems unlikely that such churn would use up all available type IDs, but if we start to run low, we could (1) add a leading version field to types that seem unstable, and possibly (2) implement custom handling in `write_object` / `read_object`, a la [`CustomTypesRegistry` in `paquito`](https://github.com/Shopify/paquito/blob/7b97f481a3690fe5531aebe2273b779481cfa4ac/lib/paquito/types.rb#L22).

2. In a `recursive: true` unpacker, we might be able to branch on a field type.  For example, if we needed to support an additional boolean flag for `Range`, we could change the last field from a boolean to a bit flag int in the packer, but still check for a boolean (and coerce it to a bit flag int) in the unpacker.

3. Alternatively, in a `recursive: true` unpacker, we might be able to add a distinct leading type and branch on that.  For example, if we know the first field in an old schema was always numeric, we could add a leading string and branch on that.

And any of the above would be conditionally enabled with a `config.*` setting for rolling deploys.

Reply:
>  think a single byte version number might make that more difficult.

I wasn't suggesting single byte, you can have a prefix that doesn't conflict with Marshal nor JSON, and is two bytes but a single integer, e.g:

```ruby
>> MessagePack.dump(128)
=> "\xCC\x80"
```

Reply:
Ah, I misunderstood!  Yes, that is a good suggestion.